Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,339 @@
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field
+`A`, then we can extend that embedding to an embedding on the completion `v.completion →+* A`. -/
+def extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    v.completion →+* L :=
+  UniformSpace.Completion.extensionHom _
+    (WithAbs.uniformInducing_of_comp h).uniformContinuous.continuous
+
+theorem extensionEmbedding_of_comp_coe
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) (x : K) :
+    extensionEmbedding_of_comp h x = f x := by
+  rw [← UniformSpace.Completion.extensionHom_coe f
+    (WithAbs.uniformInducing_of_comp h).uniformContinuous.continuous]
+  rfl
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` preserves distances. -/
+theorem extensionEmbedding_dist_eq_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)
+    (x y : v.completion) :
+    dist (extensionEmbedding_of_comp h x) (extensionEmbedding_of_comp h y) =
+      dist x y := by
+  refine (UniformSpace.Completion.induction_on₂ x y ?_ (fun x y => ?_))
+  · refine isClosed_eq ?_ continuous_dist
+    exact (continuous_iff_continuous_dist.1 (UniformSpace.Completion.continuous_extension))
+  · simp only [extensionEmbedding_of_comp_coe]
+    exact UniformSpace.Completion.dist_eq x y ▸ Isometry.dist_eq (WithAbs.isometry_of_comp h) _ _
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` is an isometry. -/
+theorem isometry_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    Isometry (extensionEmbedding_of_comp h) :=
+  Isometry.of_dist_eq <| extensionEmbedding_dist_eq_of_comp h
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` is a closed embedding. -/
+theorem closedEmbedding_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    ClosedEmbedding (extensionEmbedding_of_comp h) :=
+  (isometry_extensionEmbedding_of_comp h).closedEmbedding
+
+/-- If the absolute value of a normed field factors through an embedding into another normed and
+locally compact field, then the completion of the first normed field is also locally compact. -/
+theorem locallyCompactSpace [LocallyCompactSpace L]
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)  :
+    LocallyCompactSpace (v.completion) :=
+  (closedEmbedding_extensionEmbedding_of_comp h).locallyCompactSpace
+
+end AbsoluteValue.Completion
+
+namespace NumberField.InfinitePlace
+
+open AbsoluteValue.Completion
+
+variable {K : Type*} [Field K] (v : InfinitePlace K)
```

Comment:
I believe that these result really belong to the main file concerning infinite places, they do not have much to do with completions, no?

Reply:
Agreed, I moved them.